src/Workspaces/Remote/Core/SolutionAssetStorage.Scope.cs

```diff
@@ -53,43 +52,24 @@ public void Dispose()
         /// </summary>
         public async Task AddAssetsAsync(
             ImmutableArray<Checksum> checksums,
-            Dictionary<Checksum, SolutionAsset> assetMap,
+            Dictionary<Checksum, object> assetMap,
             CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            using var checksumsToFind = Creator.CreateChecksumSet(checksums);
+            using var obj = Creator.CreateChecksumSet(checksums);
+            var checksumsToFind = obj.Object;
 
-            var numberOfChecksumsToSearch = checksumsToFind.Object.Count;
+            var numberOfChecksumsToSearch = checksumsToFind.Count;
+            Contract.ThrowIfTrue(checksumsToFind.Contains(Checksum.Null));
 
-            if (checksumsToFind.Object.Remove(Checksum.Null))
-                assetMap[Checksum.Null] = SolutionAsset.Null;
+            await FindAssetsAsync(checksumsToFind, assetMap, cancellationToken).ConfigureAwait(false);
 
-            await FindAssetsAsync(checksumsToFind.Object, assetMap, cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfTrue(checksumsToFind.Count > 0);
             Contract.ThrowIfTrue(assetMap.Count != numberOfChecksumsToSearch);
-
-            // no checksum left to find
-            Debug.Assert(checksumsToFind.Object.Count == 0);
-        }
-
-        /// <summary>
-        /// Find an assets of the specified <paramref name="remainingChecksumsToFind"/> within <see
-        /// langword="this"/>. Once an asset of given checksum is found the corresponding asset is placed to
-        /// <paramref name="result"/> and the checksum is removed from <paramref name="remainingChecksumsToFind"/>.
-        /// </summary>
-        private async Task FindAssetsAsync(HashSet<Checksum> remainingChecksumsToFind, Dictionary<Checksum, SolutionAsset> result, CancellationToken cancellationToken)
-        {
-            using var resultPool = Creator.CreateResultSet();
-
-            await FindAssetsAsync(remainingChecksumsToFind, resultPool.Object, cancellationToken).ConfigureAwait(false);
-
-            foreach (var (checksum, value) in resultPool.Object)
-            {
-                result[checksum] = new SolutionAsset(checksum, value);
```

Comment:
no need to lookup everything first as objects, then wrap as SolutinoAssets.  all that wrapping/checking is now in the test layer only.